### DISCUSSION TITLE
How to repeat a trial (so that I can get statistical validity?)

### DISCUSSION THREAD

**Discussion by beyarkay** [optuna/optuna#4616] [Q&A] [answered]
Hi there!

I'm loving the framework, but I need to repeat certain trials so I can get an idea of the mean/variance of the performance.

For example, if a trial suggests a certain set of parameters, I want to optimize on those exact same parameters 30 times over, so that I can be sure that the results I'm getting are statistically significant and not just a byproduct of once-off randomness.

How would I do this?

My code is pretty vanilla, looking something like:
```py
study.optimize(
    lambda trial: objective(trial, X_trn, y_trn, X_val, y_val),
    n_trials=100,
    gc_after_trial=True,
)

def objective(trial, X_trn, y_trn, X_val, y_val):
    config = {
            ...
            "learning_rate": trial.suggest_float("learning_rate", 1e-6, 1e-1, log=True),
            ...
    }
    model = make_model_from_config(config=config)
    model.fit(X_trn, y_trn, validation_data=(X_val, y_val))

    final_loss = model.history.history["val_loss"][-1]
    return final_loss
```

**Comment by nzw0301** [comment, marked as answer]
I think the easiest way is to evaluate 30 times in an objective:

```python
study.optimize(
    lambda trial: objective(trial, X_trn, y_trn, X_val, y_val),
    n_trials=100,
    gc_after_trial=True,
)

def objective(trial, X_trn, y_trn, X_val, y_val):
    config = {
            ...
            "learning_rate": trial.suggest_float("learning_rate", 1e-6, 1e-1, log=True),
            ...
    }
    num_repeats = 30
    losses = []
    for _ in range(num_repeats):
        # NOTE: we need to specify different random seed values for `make_model_from_config` or `fit`
        model = make_model_from_config(config=config)
        model.fit(X_trn, y_trn, validation_data=(X_val, y_val))
        
        losses.append(model.history.history["val_loss"][-1])
    
    final_loss = np.mean(losses) # or median, etc
    return final_loss
```

We can also use [k-fold cross validation](https://scikit-learn.org/stable/modules/cross_validation.html) to create train/val datasets for each internal loop rather than using the same train/val split as in the code .

Note that we cannot use the pruning feature of Optuna for internal loop case simply.

**Reply by beyarkay** [reply]
Okay thanks. Is there any support via optuna-dashboard to display all 30 repetitions, instead of displaying just the mean of the 30 repetitions?

**Reply by nzw0301** [reply]
Sorry, I don't know about optuna-dashboard...
